src/astarte-charts/providers/deviceDatastreamObject.ts

```diff
@@ -0,0 +1,76 @@
+/*
+   This file is part of Astarte.
+
+   Copyright 2020 Ispirata Srl
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+import AstarteClient from 'astarte-client';
+import type { AstarteDatastreamObjectData } from 'astarte-client';
+
+import {
+  AstarteChartProvider,
+  AstarteChartValueObjectTimestamped,
+  AstarteChartListValue,
+} from '../index';
+
+type AstarteDeviceDatastreamObjectProviderValue = AstarteChartListValue<
+  AstarteChartValueObjectTimestamped
+>;
+
+interface AstarteDeviceDatastreamObjectProviderParams {
+  name?: string;
+  deviceId: string;
+  interfaceName: string;
+  endpoint: string;
+}
+
+export class AstarteDeviceDatastreamObjectProvider
```

Comment:
Why implementing all providers manually? the core concept of a provider is to retrieve data, and it doesn't change based on the data we're going to retrieve

```
BasicValues =
  | IndividualDatastrealmValue
  | IndividualDatastrealmValue[]
  | AggregatedDatastrealmValue
  | AggregatedDatastrealmValue[]
  | PropertyValue

class Provider<T> {
  getData: () => Promise<T> =
    () => {
      // do the api query
      // parse data
      return parsedData;
    } 
}
```

check  [this](https://stackoverflow.com/questions/40171533/typescript-call-static-method-of-generic-type) for a feel on how that's done in Typescript

Reply:
the same goes for other providers, like a possible GPS provider or device connection status provider

you can further formalize this with an abstact class `Provideable` and then having `Provider<T implements Provideable>`

Reply:
Thanks for the advice. I reworked the code so that we have a base ChartProvider class that abstracts the core concept of fetching data. Typescript's inference does its magic and now entities (providers, data, etc.) are well-typed and strictly checked.